src/Workspaces/Core/Portable/Diagnostics/Extensions.cs

```diff
@@ -372,62 +373,68 @@ private static string GetAssemblyQualifiedName(Type type)
         var analyzers = documentAnalysisScope?.Analyzers ?? compilationWithAnalyzers.Analyzers;
         var suppressionAnalyzer = analyzers.OfType<IPragmaSuppressionsAnalyzer>().FirstOrDefault();
         if (suppressionAnalyzer == null)
-        {
             return [];
-        }
 
         if (documentAnalysisScope != null)
         {
             if (documentAnalysisScope.TextDocument is not Document document)
-            {
                 return [];
-            }
 
             using var _ = ArrayBuilder<Diagnostic>.GetInstance(out var diagnosticsBuilder);
-            await AnalyzeDocumentAsync(suppressionAnalyzer, document, documentAnalysisScope.Span, diagnosticsBuilder.Add).ConfigureAwait(false);
+            await AnalyzeDocumentAsync(
+                compilationWithAnalyzers, analyzerInfoCache, suppressionAnalyzer,
+                document, documentAnalysisScope.Span, diagnosticsBuilder.Add, cancellationToken).ConfigureAwait(false);
             return diagnosticsBuilder.ToImmutableAndClear();
         }
         else
         {
             if (compilationWithAnalyzers.AnalysisOptions.ConcurrentAnalysis)
             {
-                var bag = new ConcurrentBag<Diagnostic>();
-                using var _ = ArrayBuilder<Task>.GetInstance(project.DocumentIds.Count, out var tasks);
-                foreach (var document in project.Documents)
-                {
-                    tasks.Add(AnalyzeDocumentAsync(suppressionAnalyzer, document, span: null, bag.Add));
-                }
-
-                foreach (var document in await project.GetSourceGeneratedDocumentsAsync(cancellationToken).ConfigureAwait(false))
-                {
-                    tasks.Add(AnalyzeDocumentAsync(suppressionAnalyzer, document, span: null, bag.Add));
-                }
+                using var _ = ArrayBuilder<Diagnostic>.GetInstance(out var diagnostics);
+                await ProducerConsumer<Diagnostic>.RunParallelAsync(
+                    source: project.GetAllRegularAndSourceGeneratedDocumentsAsync(cancellationToken),
```

Comment:
>GetAllRegularAndSourceGeneratedDocumentsAsync [](http://example.com/codeflow?start=36&length=45)

Why doesn't the code use project.Documents as part of it's source input either here or in the else?

Reply:
yep, this is better.